### ISSUE TITLE
Memory leaks with netcdfs + tf.data.Dataset

### ISSUE THREAD

**Issue by nbren12** [ai2cm/fv3net#1710]
@frodre and I have both been having memory leaks while training with tf datasets. If I had to guess, this performance regression is probably caused by one of the version updates from 97d2e459d917dc7d879c5405bd830b82ac8b1b82. 

We have tried a few different things:

- decrease shuffle buffer size. tf.data.Dataset doesn't seem to deallocate these buffers. This improved initial performance but memory is still leaking.
- install libtcmalloc-minimal4. Some github issues suggested this worked. I tried it here: 201edd66a7a41eb9eb0de905544efa065b7de92e. This has had the biggest impact. RNNs complete 20/25 epochs, but they still ultimately crashed.
- Add a tensorflow callback which garbage collects every 100 batches. 067f22b32221563bd8d5b268a9c2a7fc1a644ca9. This did not solve the leak.

Some ideas we've discussed trying

- [ ] use `.map` rather than `.from_generator`. We can leverage the `dict_dataset` code here. @frodre 
- [ ] remove `.prefetch`. this may hold onto old python objects. @frodre 
- [ ] see if problem persists with tfrecord data. @nbren12 

cc @mcgibbon

**Comment by AnnaKwa** [comment]
Memory usage during the nudged-to-PIRE training with the full train dataset
Left: running the  on the PR branch using tf dataset, OOMed
Right: running on master branch

<img width="587" alt="image" src="https://user-images.githubusercontent.com/16710132/159063177-e30fd9f4-4a8e-4533-be07-dd3ed2bf8a7b.png">

**Comment by mcgibbon** [comment]
This problem doesn't exist in `fv3fit.train` with `in_memory=False` when using the batches_from_netcdf loader with one batch. Here are profile reports for 3 epochs of training:
[1_epoch-2022-03-18T11:32:18.199.zip](https://github.com/ai2cm/fv3net/files/8306918/1_epoch-2022-03-18T11.32.18.199.zip)
And one epoch of training:
[3_epochs-2022-03-18T11:26:19.184.zip](https://github.com/ai2cm/fv3net/files/8306920/3_epochs-2022-03-18T11.26.19.184.zip)

The memory usage is identical for 1 or 3 epochs, and in the 3 epoch training report most of the memory allocation happens before the training function is called. The memory allocated during the training function execution (208MB) is much less than the memory allocated when seq_to_tfdataset is called (609MB). This report is from a dataset which happens to have one batch.

Two differences in that branch in fv3fit.train that could account for this are:

- I use a generator of Mapping[str, tf.Tensor] instead of a generator of tf.data.Dataset which gets flat_map'd out to a Mapping[str, tf.Tensor] later
- I don't call .prefetch in `fv3fit.train`

I'll re-run this with a data configuration that has more batches, and with in_memory=True.

**Comment by nbren12** [comment]
Here are the memory usage curves for my experiments which blew up:
https://wandb.ai/ai2cm/microphysics-emulation/reports/Untitled-Report--VmlldzoxNzA2ODQy.

**Comment by frodre** [comment]
Removing the prefetch from earlier in the chain improved memory usage.  

Dense Training
- 250 files
- 2 epochs

* **orig** - original shuffle buffers
* **smaller shuffle** - no initial buffer for creating a sample dataset, adjust training shuffle buffer to 13824
* **smaller shuffle - no prefetch** - take out prefetching from smaller shuffle experiment
* **smaller shuffle before prefetch** - move prefetching to after the shuffle call in the smaller shuffle experiment

![image](https://user-images.githubusercontent.com/1440875/159068636-e84ac8ee-df4c-459a-9204-7dbc26b50ac0.png)

**Comment by mcgibbon** [comment]
`fv3fit.train` not having this issue is robust to splitting up the batches (I split the same data into 4 batches) and setting in_memory=True and using the fv3fit.train netcdf caching layer. Same memory usage as before.

[3_epochs-4_batches-in_memory-2022-03-18T12:13:10.783.zip](https://github.com/ai2cm/fv3net/files/8307056/3_epochs-4_batches-in_memory-2022-03-18T12.13.10.783.zip)

**Comment by nbren12** [comment]
FWIW. using `dict_dataset.netcdf_url_to_dataset` which uses `.map` rather than `tf.data.Dataset.from_generator` is about 10x faster, but seems to use a bit more memory:

![out](https://user-images.githubusercontent.com/1386642/159076193-41a04ef4-0895-44c3-8bd9-221ca24f7bef.png)
blue curve is the current implementation and black is using the code from `.dict_dataset`.

Here is the code

```
from fv3fit.train_microphysics import TrainConfig
from fv3fit.emulation.data.dict_dataset import netcdf_url_to_dataset
import sys
import gc


variables = ["air_temperature_input", "specific_humidity_input"]

config = TrainConfig.from_yaml_path("projects/microphysics/train/limited.yaml")
nfiles = 60

def from_generator():
    ds = config.open_dataset(config.train_url, nfiles=nfiles, required_variables=variables)
    return ds


def dict_dataset():
    return netcdf_url_to_dataset(config.train_url, variables, nfiles=nfiles)


def run(ds):
    ds = ds.unbatch().shuffle(100_000).batch(1024)
    for i in range(10):
        gc.collect()
        print(i)
        for it in ds:
            pass


# ds = open_dict_dataset()
import sys
func_name = sys.argv[1]
ds_func = globals()[func_name]
ds = ds_func()
run(ds)

```

**Comment by nbren12** [comment]
@frodre Are you going to PR that `prefetch` order change?

**Comment by frodre** [comment]
Much slower memory leak, but still OOM'ing.  We might have to switch away from the generator.

[wandb](https://wandb.ai/ai2cm/microphysics-emulation/runs/1z1dbsq0/system?workspace=user-aperkins)

![image](https://user-images.githubusercontent.com/1440875/159136303-b97371a4-1ae1-4139-9bd3-67c469b4d8f2.png)

**Comment by nbren12** [comment]
bummer. I have several changes on a branch which might be helpful. I think we need a way to test for this bug that doesn't take 400 minutes.

**Comment by mcgibbon** [comment]
Since we don't have memory leaks on the fv3fit.train side, I looked through the emulation data loading for possible causes.

TL;DR: the biggest difference I can see is that the emulation code performs xarray operations within the generator function used to construct its dataset, and fv3fit.train does not. There are a couple other small differences.

Randomization of file order in `emulation.data.load` can have some strange behaviors as it's written. You should probably update
```python
        if random_state is None:
            random_state = np.random.RandomState(np.random.get_state()[1][0])
```
to
```python
        if random_state is None:
            random_state = np.random
```
Not sure how tensorflow parses a RandomState initialization, but at the least the way it's currently done will cause nc_dir_to_tfdataset to do a random selection without updating numpy's random state. In other words, if tensorflow calls nc_dir_to_tfdataset multiple times with nothing calling np.random in between those times, it will get an identically-ordered dataset each time. I've taken to avoiding RandomState objects entirely because of this kind of bug.

nc_files_to_tf_dataset does a very early prefetch, which we aren't doing in `fv3fit.train`. The lead author of the tf.data paper recommends [using prefetch as a last step](https://docs.google.com/presentation/d/16kHNtQslt-yuJ3w8GIx-eEH6t_AvFeQOchqGRFpAD7U/edit#slide=id.g254d08e080_0_10), e.g. to preload the next batch while training on the current batch. But while we're debugging this you might want to and might have already removed these prefetches entirely.

Noting that `self.transform.get_pipeline(required_variables)` makes it a little harder for me to tell what code is getting called here because it's dynamically defined based on config I don't have. I might have missed something.

The emulation data loading pipeline appears to have xarray transforms within the generator used to construct the `tf.data.Dataset.from_generator`. In fv3fit.train, these transforms are all either pre-computed (we then make a generator over a sequence of tf.dataset), or avoided (in the case of caching netcdfs to disk). Not sure if having xarray operations in the generator might be causing the memory leak. `netcdf_url_to_dataset` has this same behavior, where DerivedMapping is doing xarray transforms. Could test if the memory leak persists with netcdf_url_to_dataset when you have no derived variables, could even comment out the DerivedMapping call in read_variables_greedily_as_tuple.

P.S. / Note for later: This is identical in fv3fit.train and unrelated to using a RandomState, but I suspect the way the emulation code is written, the generator will return the files in the same order every time, instead of randomizing the file order each epoch. A little hard to tell for sure because the shuffling is buried deeply into the code (as opposed to something like dataset = dataset.shuffle()), but I think the filename shuffie is called before the generator is made, not within it. We could fix this once we no longer have memory issues.

**Comment by nbren12** [comment]
> Since we don't have memory leaks on the fv3fit.train side, I looked through the emulation data loading for possible causes.

Is this also the case when you train without the `in_memory` option? Anna showed the `fv3fit.train` does leak when `in_memory=False`: https://github.com/ai2cm/fv3net/issues/1710#issuecomment-1072694038. Did you fix this problem?

**Comment by mcgibbon** [comment]
Yes, it is the case with or without the in_memory option. Anna's comment earlier was not about a memory leak, the out-of-memory error occurred when loading the full dataset into memory to fit normalization statistics, before keras' `model.fit` was called. That's been fixed.

**Comment by nbren12** [comment]
And replying to your helpful comments @mcgibbon:

> Since we don't have memory leaks on the fv3fit.train side, I looked through the emulation data loading for possible causes.

Still not convinced of this. If this is true, then I think your PR should be fine to merge.

> TL;DR: the biggest difference I can see is that the emulation code performs xarray operations within the generator function used to construct its dataset, and fv3fit.train does not. There are a couple other small differences.

I agree this is a likely explanation.

> In other words, if tensorflow calls nc_dir_to_tfdataset multiple times with nothing calling np.random in between those times, it will get an identically-ordered dataset each time. I've taken to avoiding RandomState objects entirely because of this kind of bug.

This seems unrelated to memory errors and was intentional. I wanted to make sure we load the same set of files when only loading a subset of them.

> nc_files_to_tf_dataset does a very early prefetch, which we aren't doing in `fv3fit.train`. The lead author of the tf.data paper recommends [using prefetch as a last step]

1b2bfa9dac352db13a2ea476dd26ac60537a1a66 fixed this...still OOMs but much more slowly.


> Could test if the memory leak persists with netcdf_url_to_dataset when you have no derived variables, could even comment out the DerivedMapping call in read_variables_greedily_as_tuple.

Worth trying.
